Replace the webpack+acorn transform with a Babel plugin

nicolo-ribaudo opened the pull request:
I was playing around with the PDF.js build process, and I noticed that webpack runs both some Babel plugins, and a custom acorn-based transform (`pdfjsdev-loader`).

I tried to merge the Babel and the Acorn passes, to skip a AST->string->AST round-trip. With the two separate transforms this is what happens:
1. Babel parses the code
2. Babel transforms the AST
3. Babel generates the code
4. Acorn parses the code
5. pdfjsdev-loader transforms the AST
6. `@javascript-obfuscator/escodegen` generates the code
7. Webpack parses the file
8. Webpack concatenates the files

With this PR, this is reduced to:
1. Babel parses the code
2. Babel transforms the AST
3. babel-plugin-pdfjs-preprocessor transforms the AST
4. Babel generates the code
5. Webpack parses the file
6. Webpack concatenates the files

On my machine, this gives a consistent speedup in build times. `gulp lib` went **down from 3.4s to 2.6s** and `gulp dist` went **down from 36s to 29s**. Also, the Babel plugin is smaller than the acorn transforms because Babel provides two utilities that the plugin can heavily rely on:
- `t.valueToNode(val)`, to convert a JS value into the equivalent AST node. This is used when inlining JSON in the code.
- `path.evaluate()`, to statically evaluate simple expressions. This is used when simplifying the various build-time conditions to strip away unused code.

The transform was tested, so we know that the new plugin keeps the old behavior with a couple differences:
- Babel formats code differently from `@javascript-obfuscator/escodegen`, so there are some whitespaces/quotes differences. This doesn't matter in practice, as when running the transform in the full build process and not in isolation Babel was already part of the process, so it was already affecting code formatting.
- In `external/builder/fixtures_esprima/constants-expected.js` there is a `var m = '1' === true;` that is now transformed to `var m = false;`. This is ok because the runtime behavior is the same, and keeping the old behavior would require more work because we cannot use `path.evaluate()` directly.

Also, I didn't implement the `saveComments` option of `pdfjsdev-loader`. That option's only working (and used) value was `false`, because `acorn` already discards all the comments by default so there was never any comment attached to the AST.

I opened this PR because I saw that you already use a custom Babel plugin (`babelPluginReplaceNonWebpackImport`), but it doesn't actually affect users of pdf.js (it only speeds up the build for devs that integrate in in their app), so I won't be offended if you tell me that from a maintenance perspective the current setup is better 😬 However, this PR also has the advantage of removing some deps withuot introducing new ones.

nicolo-ribaudo commented:
Oh I forgot to mention that this makes the error messages when using `PDFJSDev.*` improperly slightly better. For example, for `PDFJSDev.eval("BUNDLE_VERSION FOO")`, the old error is
<img width="1157" alt="image" src="https://github.com/mozilla/pdf.js/assets/7000710/bc5e8681-7837-4215-be12-e24fed63c3e4">
while the new error is
<img width="1316" alt="image" src="https://github.com/mozilla/pdf.js/assets/7000710/ad3355a9-e9a8-4309-8fbd-a24fb149235b">

Snuffleupagus reviewed (changes requested):
Thanks for the patch; I've not actually tested this locally, but based on a very quick look the following are some initial questions/observations.
 - Please make sure that the *commit message* contains (all) relevant context from https://github.com/mozilla/pdf.js/pull/17563#issue-2093866558, such that it's possible to understand what a patch does and *importantly why* e.g. on the `git` command line as well.
 - Please squash commits, since we don't use separate fixup commits in this project.
 - Since you only mention `gulp lib`, please keep in mind that e.g. the `gulp generic` and `gulp mozcentral` builds are much more important overall (note that `gulp lib` is only used to allow unit-testing in Node.js).
 - Have you checked that this passes *all* tests, by successfully running `gulp test` (or at least `gulp test --noChrome`) locally?
   See https://github.com/mozilla/pdf.js/wiki/Contributing#4-run-lint-and-testing and note e.g. the "Generating reference images"-section (needs to be done against the `master` branch *before* testing your patch.)
 - How does this affect the final size of the builds, in particular the output of `gulp generic` *and* `gulp mozcentral`, when compared to the `master` branch?

nicolo-ribaudo commented:
(squashing now)

> * Since you only mention `gulp lib`, please keep in mind that e.g. the `gulp generic` and `gulp mozcentral` builds are much more important overall (note that `gulp lib` is only used to allow unit-testing in Node.js).

- `gulp generic` goes from 5.5s to 4.0s
- `gulp mozcentral` goes from 4.7s to 3.2s

> 
> * Have you checked that this passes _all_ tests, by successfully running `gulp test` (or at least `gulp test --noChrome`) locally?
>       See [Wiki: Contributing (4 run lint and testing) ()](https://github.com/mozilla/pdf.js/wiki/Contributing#4-run-lint-and-testing) and note e.g. the "Generating reference images"-section (needs to be done against the `master` branch _before_ testing your patch.)

Yes :) But only with `--noChrome` because Chrome tests for me are failing with this error that I haven't figured out yet how to solve: `Error while starting chrome: Network.enable timed out. Increase the 'protocolTimeout' setting in launch/connect calls for a higher timeout if needed.`.

> * How does this affect the final size of the builds, in particular the output of `gulp generic` _and_ `gulp mozcentral`, when compared to the `master` branch?

- `gulp generic`
  - `pdf.sandbox.mjs`: 612 KiB to 614 KiB (this seems to be due `@javascript-obfuscator/escodegen` being smarter than Babel when deciding aboutsingle vs double quotes for strings to avoid unnecessary escapes in the inlined `PDF_SCRIPTING_JS_SOURCE`)
  - `pdf.mjs`: 597 KiB (no change)
  - `pdf.worker.mjs`: 1.79 MiB (no change)
  - `viewer.mjs`: 510 KiB (no change)
- `gulp mozcentral` has no size change in any output file

nicolo-ribaudo committed:
Replace the webpack+acorn transform with a Babel plugin

This commit converts the pdfjsdev-loader transform into a Babel plugin,
to skip a AST->string->AST round-trip.

Before this commit, the webpack build process was:
1. Babel parses the code
2. Babel transforms the AST
3. Babel generates the code
4. Acorn parses the code
5. pdfjsdev-loader transforms the AST
6. @javascript-obfuscator/escodegen generates the code
7. Webpack parses the file
8. Webpack concatenates the files

After this commit, it is reduced to:
1. Babel parses the code
2. Babel transforms the AST
3. babel-plugin-pdfjs-preprocessor transforms the AST
4. Babel generates the code
5. Webpack parses the file
6. Webpack concatenates the files

This change improves the build time by ~25% (tested on MacBook Air M2):
- `gulp lib`: 3.4s to 2.6s
- `gulp dist`: 36s to 29s
- `gulp generic`: 5.5s to 4.0s
- `gulp mozcentral`: 4.7s to 3.2s

The new Babel plugin doesn't support the `saveComments` option of
pdfjsdev-loader, and it just always discards comments. Even though
pdfjsdev-loader supported multiple values for that option, it was
effectively ignored due to `acorn` dropping comments by default.

nicolo-ribaudo commented:
@Snuffleupagus I'm not sure about what's going on with CI -- could you try restarting it?

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus commented:
/botio test

nicolo-ribaudo commented:
Uhm I'm not sure why I did't catch those differences locally (not even the firefox-related ones). I'll try diffing the output bundles to see what could cause it.

nicolo-ribaudo commented:
Or is it expected that sometimes there are some slight variations in color of some pixels? e.g.
<img width="337" alt="image" src="https://github.com/mozilla/pdf.js/assets/7000710/578c1758-f7e2-408b-b877-26f5072e3e5b">

I see that some PRs are merged with this type of failures.

Snuffleupagus commented:
Given that Firefox was just updated to version 124, most of the movement is probably related to that. I'll trigger testing and makeref in a PR without any code changes, and then we can re-run tests here.

Unfortunately, as you've already noticed, we also have a number of intermittent failures with *small* pixel differences in the render. While obviously annoying, this is a known issue that's nothing to worry about.

---

However, having just tested this locally and diffed the output of `gulp mozcentral` for the `master` branch and this patch there's now some unnecessary `return;`-lines included in the output; see an example diff below.

For example, [this method](https://github.com/mozilla/pdf.js/blob/8b247221138fb8885e1e48a18ea1a6257eb7b560/web/view_history.js#L58-L66) once built with `gulp mozcentral` now includes a "useless" `return;` which wasn't previously the case.

```diff
--- C:\Users\Jonas\Desktop\mozcentral\browser\extensions\pdfjs\content\web\viewer-geckoview.mjs
+++ C:\Users\Jonas\Git\pdf.js\build\mozcentral\browser\extensions\pdfjs\content\web\viewer-geckoview.mjs
@@ -5030,7 +5030,7 @@
   #scaleTimeoutId = null;
   #textLayerMode = TextLayerMode.ENABLE;
   constructor(options) {
-    const viewerVersion = '4.1.75';
+    const viewerVersion = "4.1.76";
     if (version !== viewerVersion) {
       throw new Error(`The API version "${version}" does not match the Viewer version "${viewerVersion}".`);
     }
@@ -6103,7 +6103,9 @@
   get scrollMode() {
     return this._scrollMode;
   }
-  set scrollMode(mode) {}
+  set scrollMode(mode) {
+    return;
+  }
   _updateScrollMode(pageNumber = null) {
     const scrollMode = this._scrollMode,
       viewer = this.viewer;
@@ -6128,7 +6130,9 @@
   get spreadMode() {
     return this._spreadMode;
   }
-  set spreadMode(mode) {}
+  set spreadMode(mode) {
+    return;
+  }
   _updateSpreadMode(pageNumber = null) {
     if (!this.pdfDocument) {
       return;
@@ -6509,6 +6513,7 @@
   async _writeToStorage() {
     const databaseStr = JSON.stringify(this.database);
     sessionStorage.setItem("pdfjs.history", databaseStr);
+    return;
   }
   async _readFromStorage() {
     return sessionStorage.getItem("pdfjs.history");
@@ -7528,7 +7533,9 @@
       source: this
     });
   },
-  async _initializePageLabels(pdfDocument) {},
+  async _initializePageLabels(pdfDocument) {
+    return;
+  },
   _initializePdfHistory({
     fingerprint,
     viewOnLoad,
@@ -8610,34 +8617,34 @@
 
 class BasePreferences {
   #defaults = Object.freeze({
-    "annotationEditorMode": 0,
-    "annotationMode": 2,
-    "cursorToolOnLoad": 0,
-    "defaultZoomDelay": 400,
-    "defaultZoomValue": "",
-    "disablePageLabels": false,
-    "enableHighlightEditor": false,
-    "enablePermissions": false,
-    "enablePrintAutoRotate": true,
-    "enableScripting": true,
-    "externalLinkTarget": 0,
-    "highlightEditorColors": "yellow=#FFFF98,green=#53FFBC,blue=#80EBFF,pink=#FFCBE6,red=#FF4F5F",
-    "historyUpdateUrl": false,
-    "ignoreDestinationZoom": false,
-    "forcePageColors": false,
-    "pageColorsBackground": "Canvas",
-    "pageColorsForeground": "CanvasText",
-    "pdfBugEnabled": false,
-    "sidebarViewOnLoad": -1,
-    "scrollModeOnLoad": -1,
-    "spreadModeOnLoad": -1,
-    "textLayerMode": 1,
-    "viewOnLoad": 0,
-    "disableAutoFetch": false,
-    "disableFontFace": false,
-    "disableRange": false,
-    "disableStream": false,
-    "enableXfa": true
+    annotationEditorMode: 0,
+    annotationMode: 2,
+    cursorToolOnLoad: 0,
+    defaultZoomDelay: 400,
+    defaultZoomValue: "",
+    disablePageLabels: false,
+    enableHighlightEditor: false,
+    enablePermissions: false,
+    enablePrintAutoRotate: true,
+    enableScripting: true,
+    externalLinkTarget: 0,
+    highlightEditorColors: "yellow=#FFFF98,green=#53FFBC,blue=#80EBFF,pink=#FFCBE6,red=#FF4F5F",
+    historyUpdateUrl: false,
+    ignoreDestinationZoom: false,
+    forcePageColors: false,
+    pageColorsBackground: "Canvas",
+    pageColorsForeground: "CanvasText",
+    pdfBugEnabled: false,
+    sidebarViewOnLoad: -1,
+    scrollModeOnLoad: -1,
+    spreadModeOnLoad: -1,
+    textLayerMode: 1,
+    viewOnLoad: 0,
+    disableAutoFetch: false,
+    disableFontFace: false,
+    disableRange: false,
+    disableStream: false,
+    enableXfa: true
   });
   #prefs = Object.create(null);
   #initializedPromise = null;
@@ -8650,13 +8657,13 @@
       prefs
     }) => {
       const BROWSER_PREFS = {
-        "canvasMaxAreaInBytes": -1,
-        "isInAutomation": false,
-        "supportsDocumentFonts": true,
-        "supportsIntegratedFind": false,
-        "supportsMouseWheelZoomCtrlKey": true,
-        "supportsMouseWheelZoomMetaKey": true,
-        "supportsPinchToZoom": true
+        canvasMaxAreaInBytes: -1,
+        isInAutomation: false,
+        supportsDocumentFonts: true,
+        supportsIntegratedFind: false,
+        supportsMouseWheelZoomCtrlKey: true,
+        supportsMouseWheelZoomMetaKey: true,
+        supportsPinchToZoom: true
       };
       const options = Object.create(null);
       for (const [name, defaultVal] of Object.entries(BROWSER_PREFS)) {
@@ -9242,8 +9249,8 @@
 
 
 
-const pdfjsVersion = '4.1.75';
-const pdfjsBuild = '8b2472211';
+const pdfjsVersion = "4.1.76";
+const pdfjsBuild = "7ba5cae77";
 const AppConstants = null;
 window.PDFViewerApplication = PDFViewerApplication;
 window.PDFViewerApplicationConstants = AppConstants;
```

nicolo-ribaudo commented:
Oh, fixed (diff: https://github.com/mozilla/pdf.js/compare/8fb9ba9add01b92f592dbcbc54d5e1ef77c16db1..6db7bf62fe8b5e36af1e426cfc33f7b5cd4074c6). I had forgot that acorn uses FunctionExpression also for methods.

nicolo-ribaudo commented on external/builder/test-fixtures_esprima.mjs:
```diff
@@ -48,6 +48,10 @@ files.forEach(function (expectationFilename) {
     errors++;
 
+    if (process.env.OVERWRITE) {
+      fs.writeFileSync(expectationFilename, out);
```
I committed this by accident, but it's quite handy so probably I can leave it. It makes it possible to simply run `OVERWRITE=true node ./external/builder/test-fixtures_esprima.mjs` to regenerate the expected output in case of a change.

Snuffleupagus replied:
I'm fine with doing this, with the following two things addressed:
 - Add a code comment explaining what it's for, and how to use this, to help future readers of this code actually use the new functionality.
 - Add the same functionality in https://github.com/mozilla/pdf.js/blob/master/external/builder/test-fixtures.mjs as well, for consistency across these tests.

Snuffleupagus commented:
Another question, to help both with reviewing (since it might produce a smaller diff) and with preserving blame: Would it be possible do `git mv external/builder/preprocessor2.mjs external/builder/babel-plugin-pdfjs-preprocessor.mjs`?

nicolo-ribaudo committed:
Rename preprocessor2.mjs to babel-plugin-pdfjs-preprocessor.mjs

This is in preparation for the next commit, which will convert
preprocessor2.mjs to a Babel plugin. The purpose of this commit
is to help git track the rename regardless of the large amount
of changes.

nicolo-ribaudo commented:
Done, but I had to keep it as a separate commit otherwise Git cannot track the rename.

Snuffleupagus commented on external/builder/babel-plugin-pdfjs-preprocessor.mjs:
```diff
@@ -0,0 +1,241 @@
+          throw new Error("Path to JSON is not provided");
+        }
+        let jsonPath = arg.value;
+        if (jsonPath.indexOf(ROOT_PREFIX) === 0) {
```
This is obviously a pre-existing issue, but could we change this as follows?
```suggestion
        if (jsonPath.startsWith(ROOT_PREFIX)) {
```

(We have ESLint rules that *should* catch these cases, but given that ESLint is a static code analyser I'm guessing that it cannot safely tell that `jsonPath` is a string.)

Snuffleupagus commented on external/builder/babel-plugin-pdfjs-preprocessor.mjs:
```diff
@@ -0,0 +1,241 @@
+    },
+    visitor: {
+      "ExportNamedDeclaration|ImportDeclaration": ({ node }) => {
+        if (node.source && ctx.map && ctx.map[node.source.value]) {
```
Looking at the old code, is it correct/safe to drop the `node.source.type === "Literal"` check? (I'm not that knowledgeable about Babel, hence why I wanted to ask.)

Also, we might take the opportunity to "modernize" this a little bit.
```suggestion
        if (node.source && ctx.map?.[node.source.value]) {
```

nicolo-ribaudo replied:
It's safe because the source of an import/export is always a string. I think this check was accidentally copyied over from the `require(...)` case, where the argument could be anything.

Snuffleupagus commented on external/builder/babel-plugin-pdfjs-preprocessor.mjs:
```diff
@@ -0,0 +1,241 @@
+            case "===":
+            case "!=":
+            case "!==":
+              // folding == and != check that can be statically evaluated
+              const { confident, value } = path.evaluate();
```
How "confident", pun intended, can we be about this *always* doing the right thing?

Some quick searching brought up https://github.com/babel/babel/issues/14197, which suggests that there *may* be problems in some cases (unless I'm misunderstanding how this works).

Snuffleupagus commented:
/botio-linux preview

Snuffleupagus commented:
/botio test

Snuffleupagus commented:
/botio-windows test

Snuffleupagus reviewed (approved):
The build-time improvements obviously cannot hurt, and reducing the number of direct dependencies needed for building is also very good.
(Especially getting rid of `escodegen` seems beneficial since it hasn't been very "actively" maintained the last few years, which forced us to switch to a fork.)

r=me, thank you for the patch!

nicolo-ribaudo commented:
Thanks for helping and reviewing :)

Snuffleupagus commented:
@nicolo-ribaudo Unfortunately this seems to have caused a regression, see issue #17589; can you please help out with that?

nicolo-ribaudo commented:
Yes I'll take a look tomorrow or on Monday!

nicolo-ribaudo commented on external/builder/babel-plugin-pdfjs-preprocessor.mjs:
```diff
@@ -0,0 +1,241 @@
+          }
+        },
+      },
+      UnaryExpression(path) {
```
The fix is probably to add `exit` here like to the other visitors (I can submit a PR tomorrow, I'm from my phone right now), so that the UnaryExpression is checked after that the inner PDFJSDev.test call is replaced.
